### PR TITLE
[WIP] PD: Pad to multiple faces

### LINKED ISSUE #11135: [Feature Request] PartDesign: Pad UpTo multiple faces
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

PartDesign_Pad only allow to select one face to end to. In cases of the base's projection is partially out of one face it's not sufficient (see #11111 and #10538).

### Full version info

```shell
[code]
OS: Debian GNU/Linux trixie/sid (XFCE/lightdm-xsession)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.34594 (Git)
Build type: Unknown
Branch: main
Hash: 3d9e3fa809e7dbc3ec60ed23195759b5aff4ed54
Python 3.11.6, Qt 5.15.10, Coin 4.0.0, Vtk 9.1.0, OCC 7.6.3
Locale: French/France (fr_FR)
[/code]
```

### Subproject(s) affected?

PartDesign

### Anything else?

OCC take a shape as limit:
`BRepFeat_MakePrism::Perform (const TopoDS_Shape& Until)`
but we provide a face:
```
void FeatureExtrude::generatePrism(TopoDS_Shape& prism,
                                   const std::string& method,
                                   const TopoDS_Shape& baseshape,
                                   const TopoDS_Shape& profileshape,
                                   const TopoDS_Face& supportface,
                                   const TopoDS_Face& uptoface,
                                   const gp_Dir& direction,
                                   PrismMode Mode,
                                   Standard_Boolean Modify)
// […]
        BRepFeat_MakePrism PrismMaker;
// […]
            PrismMaker.Perform(uptoface);
```

We should create a TopoDS_Shape from selected faces and provide it to generatePrism.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by FlachyJoe**
Aims to fix #11135.

### CODE REVIEW

**Comment by FlachyJoe** [comment]
First working UpToShape implementation:
image

**Comment by maxwxyz** [comment]
This looks great and will also help https://github.com/FreeCAD/FreeCAD/issues/10538

**Comment by FlachyJoe** [comment]
With this small changes a new Up To Shape mode is introduced for Pad.
The prism is made up to the base shape until the end face is closed by existant ones.
It's the default OCC feature so it needed few works :smile: 
image



I have a Up To FaceS in plan but it doesn't work as expected yet.
This first step is ready to review and to test.

Already known problems:

- No midplane
- If no existant faces can completely limit the prism, it goes throw all and more.
image

File: src/Mod/PartDesign/App/FeatureExtrude.cpp [hunk 1]
```diff
@@ -194,7 +194,7 @@ void FeatureExtrude::generatePrism(TopoDS_Shape& prism,
                                    const TopoDS_Shape& baseshape,
                                    const TopoDS_Shape& profileshape,
                                    const TopoDS_Face& supportface,
-                                   const TopoDS_Face& uptoface,
+                                   const TopoDS_Shape& uptoface,
```
Comment (AjinkyaDahale) [inline src/Mod/PartDesign/App/FeatureExtrude.cpp:197]: Might want to start calling this `uptoshape`.

**Comment by maxwxyz** [comment]
does not work for the example file of https://github.com/FreeCAD/FreeCAD/issues/11111#issuecomment-1772432318 [extrude.zip](https://github.com/FreeCAD/FreeCAD/files/13328327/extrude.zip)
Where do I find the new setting?
grafik
```
OS: macOS 13.6.1
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.34982 (Git)
Build type: Release
Branch: main
Hash: 55639df57325290e5a9d98ef48d6e17898fe0fc1
Python 3.11.0, Qt 5.15.8, Coin 4.0.0, Vtk 9.2.5, OCC 7.6.3
Locale: German/Germany (de_DE)
Installed mods: 
  * 3DfindIT 1.2.0
  * freecad.gears 1.0.0
  * OpenDark 2023.10.30
  * CfdOF 1.24.8
  * Defeaturing 1.2.1
  * Manipulator 1.5.0
  * fasteners 0.4.70
  * CurvedShapes 1.0.5
  * sheetmetal 0.3.12
  * 3D_Printing_Tools
  * Curves 0.6.14
[/code]
```

**Comment by FlachyJoe** [comment]
@maxwxyz this PR is not merged yet. If you compile it then you'll see 
Capture d’écran_2023-11-12_16-45-43

**Comment by AjinkyaDahale** [comment]
> @maxwxyz this PR is not merged yet. If you compile it then you'll see Capture d’écran_2023-11-12_16-45-43

There is also the an option to create snaps if you're interested.

**Comment by maxwxyz** [comment]
I've compiled this PR. What do you mean by snaps?

**Comment by AjinkyaDahale** [comment]
> I've compiled this PR.

It looks like a straightforward PR. Could you share the about info from your compiled build?

> What do you mean by snaps?

The developer can get a snap build through https://github.com/FreeCAD/FreeCAD-snap. That's an option for Flachyjoe, however, and only useful if the user can't/doesn't want to compile and is using Linux.

**Comment by maxwxyz** [comment]
Ok maybe I've done something wrong, will compile again

**Comment by maxwxyz** [comment]
Compiled again but not visible in the task dialog: 
grafik

But visible in the data tab:
grafik


Is working!
```
[code]
OS: macOS 13.6.1
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.34982 (Git)
Build type: Release
Branch: main
Hash: 55639df57325290e5a9d98ef48d6e17898fe0fc1
Python 3.11.0, Qt 5.15.8, Coin 4.0.0, Vtk 9.2.5, OCC 7.6.3
Locale: German/Germany (de_DE)
Installed mods: 
  * 3DfindIT 1.2.0
  * freecad.gears 1.0.0
  * OpenDark 2023.10.30
  * CfdOF 1.24.8
  * Defeaturing 1.2.1
  * Manipulator 1.5.0
  * fasteners 0.4.70
  * CurvedShapes 1.0.5
  * sheetmetal 0.3.12
  * 3D_Printing_Tools
  * Curves 0.6.14
[/code]

```

**Comment by FlachyJoe** [comment]
@maxwxyz As I noted somewhere I don't plane to update the GUI and I'll open an issue for this when this PR will be merged.

**Comment by FlachyJoe** [comment]
@luzpaz I already have to wait long enough to compile on my old computer not to want to deal with snap.

**Comment by luzpaz** [comment]
@FlachyJoe Clarrification, do you want a snap built for this PR ?

**Comment by FlachyJoe** [comment]
@luzpaz I now read the snap repository doc and understand the compilation is done on the server. It could be interesting to submit the PR to many testers via the forum but I don't think this small one needs it.

**Comment by FlachyJoe** [comment]
Thank you @sliptonic

**Comment by luzpaz** [comment]
Thank you for your work on this @FlachyJoe!

**Comment by dodomarg** [comment]
Hi,
First thanks for the work with this feature! 
I believe I stumbled on an edge case that was not discussed here. Seems to be when trying to pad a concave shape above a cretin size or something else (unclear to me) the feature doesn't generate geometry. 

Example that works fine:
freecad_pad_UpToShape_6(sketch)
freecad_pad_UpToShape_6

Example that brakes down: 
freecad_pad_UpToShape_7(sketch
freecad_pad_UpToShape_7

This behavior is consistent regardless if the sketch is fully contained or not if that matters and I have also attached another file where the behavior is appearing. I am not very experienced with freecad at all so IDK if this is at the end an issue with my process.

Some system info: 
```

FreeCAD 0.22.0dev Revision: 35808 (Git)

snap                         2.61.1
snapd                        unavailable
series                       16
Windows Subsystem for Linux  -
kernel                       5.15.133.1-microsoft-standard-WSL2 (amd64)

Distributor ID: Ubuntu
Description:    Ubuntu 22.04.3 LTS
Release:        22.04
Codename:       jammy

```

[UpToShape_edge_case.zip](https://github.com/FreeCAD/FreeCAD/files/14230236/UpToShape_edge_case.zip)  - just a small test case


[filter fan - Copy.zip](https://github.com/FreeCAD/FreeCAD/files/14230246/filter.fan.-.Copy.zip)  - the original point I came across the edge case

**Comment by FlachyJoe** [comment]
@dodomarg thank you to point this out. This case will, at least, raise an error in [the next implementation I'm working on](https://github.com/FreeCAD/FreeCAD/pull/11392).

**Comment by FlachyJoe** [comment]
@bgbsww `FeatureExtrude::buildExtrusion` doesn't implement this feature. It causes a regression when `FC_USE_TNP_FIX` is activated.
#11392 will change this but isn't ready yet and I don't know if I could finalize it before feature freeze.

**Comment by bgbsww** [comment]
I'm a little confused on how to proceed here.  It looks like #11392 might be on track.

If that goes, then there are no changes required?  But if it doesn't, the code changes to buildExtrusion need to be added separately?  Can that just be a PR on it's own?

File: src/Mod/PartDesign/App/FeaturePad.cpp [hunk 2]
```diff
@@ -137,23 +142,24 @@ App::DocumentObjectExecReturn *Pad::execute()
...
-                getFaceFromLinkSub(upToFace, UpToFace);
-                upToFace.Move(invObjLoc);
+            if (method != "UpToShape") {
+                // Find a valid face or datum plane to extrude up to
+                if (method == "UpToFace") {
+                    getFaceFromLinkSub(upToFace, UpToFace);
+                    upToFace.Move(invObjLoc);
+                }
+                getUpToFace(upToFace, base, sketchshape, method, dir);
+                addOffsetToFace(upToFace, dir, Offset.getValue());
```
Comment (FlachyJoe) [inline src/Mod/PartDesign/App/FeaturePad.cpp:161]: These lines are copied at FeatureExtrude:571

File: src/Mod/PartDesign/App/FeaturePad.cpp [hunk 3]
```diff
@@ -137,23 +142,24 @@ App::DocumentObjectExecReturn *Pad::execute()
         sketchshape.Move(invObjLoc);
 
         TopoDS_Shape prism;
-        std::string method(Type.getValueAsString());
-        if (method == "UpToFirst" || method == "UpToLast" || method == "UpToFace") {
+        if (method == "UpToFirst" || method == "UpToLast" || method == "UpToFace"  || method == "UpToShape") {
```
Comment (FlachyJoe) [inline src/Mod/PartDesign/App/FeaturePad.cpp:145]: The UpToShape method lacks at FeatureExtrude:559

File: src/Mod/PartDesign/App/FeaturePad.cpp [hunk 4]
```diff
@@ -182,7 +191,10 @@ App::DocumentObjectExecReturn *Pad::execute()
                 if (!Ex.More())
                     supportface = TopoDS_Face();
                 PrismMode mode = PrismMode::None;
-                generatePrism(prism, method, base, sketchshape, supportface, upToFace, dir, mode, Standard_True);
+                if (method == "UpToShape")
+                    generatePrism(prism, "UpToFace", base, sketchshape, supportface, base, dir, mode, Standard_True);
+                else
+                    generatePrism(prism, method, base, sketchshape, supportface, upToFace, dir, mode, Standard_True);
```
Comment (FlachyJoe) [inline src/Mod/PartDesign/App/FeaturePad.cpp:197]: These lines are not in FeatureExtrude::BuildExtrusion

File: src/Mod/PartDesign/App/FeaturePad.cpp [hunk 5]
```diff
@@ -165,7 +171,10 @@ App::DocumentObjectExecReturn *Pad::execute()
                     supportface = TopoDS_Face();
 
                 PrismMode mode = PrismMode::None;
-                generatePrism(prism, method, base, sketchshape, supportface, upToFace, dir, mode, Standard_True);
+                if (method == "UpToShape")
+                    generatePrism(prism, "UpToFace", base, sketchshape, supportface, base, dir, mode, Standard_True);
+                else
+                    generatePrism(prism, method, base, sketchshape, supportface, upToFace, dir, mode, Standard_True);
```
Comment (FlachyJoe) [inline src/Mod/PartDesign/App/FeaturePad.cpp:177]: @bgbsww These lines are not in FeatureExtrude::BuildExtrusion

**Comment by FlachyJoe** [comment]
> If that goes, then there are no changes required?

Yes



> But if it doesn't, the code changes to buildExtrusion need to be added separately? Can that just be a PR on it's own?

Yes, to keep the UpTo(the base)Shape feature
